crates/cli/src/main.rs

```diff
@@ -1,52 +1,226 @@
-use anyhow::{anyhow, bail, Result};
+use anyhow::{anyhow, bail, Context, Ok, Result};
 use clap::Parser;
 use core::fmt::Write;
-use std::fs;
+use std::{
+    ffi::OsStr,
+    fs,
+    path::{Path, PathBuf},
+};
 use wasmi::{
     core::{Value, ValueType, F32, F64},
+    Engine,
     ExportItemKind,
     Func,
     FuncType,
+    Linker,
+    Module,
     Store,
 };
+use wasmi_wasi::{ambient_authority, Dir, TcpListener, WasiCtx, WasiCtxBuilder};
+
+fn parse_env(s: &str) -> Result<(String, String)> {
+    s.split_once('=')
+        .map(|(s1, s2)| (s1.to_owned(), s2.to_owned()))
+        .ok_or_else(|| anyhow::anyhow!("expects '='-separated pair"))
+}
 
 /// Simple program to greet a person
 #[derive(Parser, Debug)]
 #[clap(author, version, about, long_about = None)]
 struct Args {
-    /// The WebAssembly file to execute.
-    #[clap(value_parser)]
-    wasm_file: String,
+    /// The host directory to pre-open for the `guest` to use.
+    #[clap(long = "dir", number_of_values = 1)]
+    dirs: Vec<String>,
+
+    /// The socket address provided to the module. Allows it to perform socket-related `WASI` ops.
+    #[clap(long = "tcplisten", number_of_values = 1)]
+    tcplisten: Vec<String>,
+
+    /// The environment variable pair made available for the program.
+    #[clap(long = "env", value_parser(parse_env))]
+    envs: Vec<(String, String)>,
```

Comment:
I am not sure how to deal with mappings in `clap` but since mappings in CLI apps are ubiquitous I feel like there already is a mechanism that is more straight forward than to declare your own functionality with a `(String, String)` tuple. Correct me if I am wrong. I am just scrolling through `clap`'s manual myself.

Reply:
What we want is something like this:
```
--env myflag=foo --env otherflag=bar --env bazflag=kro
```

Reply:
> What we want is something like this:  `--env myflag=foo --env otherflag=bar --env bazflag=kro`

Yeah I will look more closely

Reply:
The `clap` manual provides an example of a definition list: https://docs.rs/clap/latest/clap/_derive/_cookbook/typed_derive/index.html#example-custom-types-derive-api with the `defines` flag:
```rust
    /// Hand-written parser for tuples
    #[arg(short = 'D', value_parser = parse_key_val::<String, i32>)]
    defines: Vec<(String, i32)>,
```
And a utility method:
```rust
/// Parse a single key-value pair
fn parse_key_val<T, U>(s: &str) -> Result<(T, U), Box<dyn Error + Send + Sync + 'static>>
where
    T: std::str::FromStr,
    T::Err: Error + Send + Sync + 'static,
    U: std::str::FromStr,
    U::Err: Error + Send + Sync + 'static,
{
    let pos = s
        .find('=')
        .ok_or_else(|| format!("invalid KEY=value: no `=` found in `{}`", s))?;
    Ok((s[..pos].parse()?, s[pos + 1..].parse()?))
}
```

Reply:
It is kinda close to what you are already doing though but maybe still a nice example for you.